server/src/handler_driver.c

```diff
@@ -168,6 +192,412 @@ static bool drive_route_params_valid(const char *train_id, t_interlocking_route
+		info_arr.len++;
+		if (info_arr.data_ptr[i] == NULL) {
+			syslog_server(LOG_ERR, 
+			              "Building signal-infos: unable to allocate memory for new element");
+			free_route_signal_info_array(&info_arr);
+			return info_arr;
+		}
+		
+		// C. Set t's trivial members to appropriate defaults (set to stop, is source, is destination)
+		info_arr.data_ptr[i]->has_been_set_to_stop = false;
+		info_arr.data_ptr[i]->is_source_signal = (i == 0);
+		info_arr.data_ptr[i]->is_destination_signal = (i+1 == number_of_signal_infos);
+		
+		// D. Allocate memory for non-trivial member 'id' in t
+		info_arr.data_ptr[i]->id = (char *) malloc(sizeof(char) * (strlen(signal_id_item) + 1));
+		if (info_arr.data_ptr[i]->id == NULL) {
+			syslog_server(LOG_ERR, "Building signal-infos: unable to allocate memory for "
+			              "new element ID");
+			free_route_signal_info_array(&info_arr);
+			return info_arr;
+		}
+		
+		// E. Copy ID from signal (from route->signals) into newly allocated 'id' of t
+		strcpy(info_arr.data_ptr[i]->id, signal_id_item);
+		
+		// F. Determine Index of the current signal in route->path and set member 
+		//    for that index of t accordingly
+		info_arr.data_ptr[i]->index_in_route_path = 0;
+		if (!info_arr.data_ptr[i]->is_source_signal && !info_arr.data_ptr[i]->is_destination_signal) {
+			for (size_t path_index = 0; path_index < route->path->len; ++path_index) {
+				const char *path_item = g_array_index(route->path, char *, path_index);
+				if (path_item != NULL && strcmp(path_item, signal_id_item) == 0) {
+					info_arr.data_ptr[i]->index_in_route_path = path_index;
+					break;
+				}
+			}
+		}
+		++built_signal_infos_counter;
+	}
+	return info_arr;
+}
+
+
+void free_multi_segment_indices(t_route_multi_segment_indices *d_seg_indices) {
+	if (d_seg_indices != NULL && d_seg_indices->arr != NULL) {
+		free(d_seg_indices->arr);
+		d_seg_indices->arr = NULL;
+		d_seg_indices->len = 0;
+	}
+}
+
+// "Multi segment index" = Index of a segment that appears more than once in a route
+static t_route_multi_segment_indices get_multi_segment_indices_for_route(const t_interlocking_route *route) {
+	// For route, determine which segments appear more than once
+	t_route_multi_segment_indices d_seg_indices = {.arr = NULL, .len = 0};
+	if (route == NULL || route->path == NULL) {
+		return d_seg_indices;
+	}
+	// First allocate enough memory for more than the maximum possible amount of 
+	// duplicate segment indices. Will afterwards be reallocated to the size it actually needs.
+	d_seg_indices.arr = malloc(sizeof(size_t) * route->path->len);
+	if (d_seg_indices.arr == NULL) {
+		syslog_server(LOG_ERR, 
+		              "Get multi segment indices for route %s: Unable to allocate memory", route->id);
+		return d_seg_indices;
+	}
+	// For every segment in route->path, check if it occurs again at a different position. 
+	// If yes, add both occurrences to d_seg_indices.
```

Comment:
The length of `d_seg_indices` may exceed `route->path->len` in some cases (see line 350).

Reply:
To first comment: The comment on lines 356, 357 is out of date. As can be seen on line 365, only one of the segments is added. This would still cause it to be added more than once if a segment appears three times. However, that can be solved with a `break` after line 366. The other occurrence of this segment will be caught once the outer loop gets there still. In turn, that would also ensure that d_seg_indices is always shorter than route->path->len.